### PR TITLE
[FPGA] Add Host Pipes Tutorial

### PR DESCRIPTION
**Pull request by rho180**
# Adding a New Sample(s)

## Description

This FPGA tutorial demonstrates how to use pipes to send and receive data between a host and a device. Pipes are a first-in first-out (FIFO) buffer construct that provide links between elements of a design.

Please include a description of the sample

## Checklist
Administrative
- [x] Review sample design with the appropriate [Domain Expert](https://github.com/oneapi-src/oneAPI-samples/wiki/Reviewers-and-Domain-Experts): <insert Name Here>
- [x] If you have any new dependencies/binaries, inform the oneAPI Code Samples Project Manager

Code Development
- [x] Implement coding guidelines and ensure code quality. [see wiki for details](https://github.com/oneapi-src/oneAPI-samples/wiki/General-Code-Guidelines)
- [x] Adhere to readme template 
- [x] Enforce format via clang-format config file
- [x] Adhere to sample.json specification. https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-specification
- [x] Ensure/create CI test configurations for sample (ciTests field) https://github.com/oneapi-src/oneAPI-samples/wiki/sample-json-ci-test-object
- [x] Run jsonlint on sample.json to verify json syntax. www.jsonlint.com

Security and Legal
- [ ] OSPDT Approval (see Project Manager for assistance)
- [ ] Compile using the following compiler flags and fix any warnings, the falgs are: "/Wall -Wformat-security -Werror=format-security"
- [ ] Bandit Scans (Python only)
- [ ] Virus scan

Review
- [ ] Review DPC++ code with Paul Peterseon. (GitHub User: pmpeter1)
- [ ] Review readme with Tom Lenth(@tomlenth) and/or Project Manager
- [ ] Tested using Dev Cloud when applicable

### CODE REVIEW

File: DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/src/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,94 @@
+# To see a Makefile equivalent of this build system:
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/src/CMakeLists.txt:1]:
```suggestion
Please remove this 3-line comment 
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/README.md [hunk 2]
```diff
@@ -0,0 +1,438 @@
+
+
+# Host Pipes
+This FPGA tutorial demonstrates how to use pipes to send and receive data between a host and a device. Pipes are a first-in first-out (FIFO) buffer construct that provide links between elements of a design. Access pipes through read and write application programming interfaces (APIs), without the notion of a memory address or pointer to elements within the FIFO. Pipes that connect a host and a device are referred to as host pipes.
+
+***Documentation***:  The [DPC++ FPGA Code Samples Guide](https://software.intel.com/content/www/us/en/develop/articles/explore-dpcpp-through-intel-fpga-code-samples.html) helps you to navigate the samples and build your knowledge of DPC++ for FPGA. <br>
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/README.md:6]:
```suggestion
***Documentation***:  The [SYCL FPGA Code Samples Guide](https://software.intel.com/content/www/us/en/develop/articles/explore-dpcpp-through-intel-fpga-code-samples.html) helps you to navigate the samples and build your knowledge of SYCL for FPGA. <br>
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/README.md [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+The [oneAPI DPC++ FPGA Optimization Guide](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide) is the reference manual for targeting FPGAs through DPC++. <br>
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/README.md:7]:
```suggestion
The [oneAPI SYCL FPGA Optimization Guide](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide) is the reference manual for targeting FPGAs through SYCL. <br>
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/README.md [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+The [oneAPI Programming Guide](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/) is a general resource for target-independent DPC++ programming.
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/README.md:8]:
```suggestion
The [oneAPI Programming Guide](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/) is a general resource for target-independent SYCL programming.
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/README.md [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+
+| Optimized for                     | Description
+---                                 |---
+| OS                                | Linux* Ubuntu* 18.04/20.04, RHEL*/CentOS* 8, SUSE* 15; Windows* 10
+| Hardware                          | Intel® FPGA Programmable Acceleration Card (PAC) D5005 (with Intel Stratix® 10 SX) <br> Intel® FPGA 3rd party / custom platforms with oneAPI support (and SYCL USM support) <br> *__Note__: Intel® FPGA PAC hardware is only compatible with Ubuntu 18.04*
+| Software                          | Intel® oneAPI DPC++ Compiler <br> Intel® FPGA Add-On for oneAPI Base Toolkit
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/README.md:14]:
```suggestion
| Software                          | Intel® oneAPI DPC++/C++ Compiler <br> Intel® FPGA Add-On for oneAPI Base Toolkit
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/README.md [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+
+#### Additional template parameters
+
+Host pipes use additional template parameters beyond the three described above. The use of these parameters is beyond the scope of this tutorial; their definitions and usage can be found in the [oneAPI DPC++ FPGA Optimization Guide](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide). Suitable values for these parameters consistent with non-specialized host pipe usage have been used in the accompanying tutorial code.
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/README.md:73]:
```suggestion
Host pipes use additional template parameters beyond the three described above. The use of these parameters is beyond the scope of this tutorial; their definitions and usage can be found in the [oneAPI SYCL FPGA Optimization Guide](https://software.intel.com/content/www/us/en/develop/documentation/oneapi-fpga-optimization-guide). Suitable values for these parameters consistent with non-specialized host pipe usage have been used in the accompanying tutorial code.
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/README.md [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+the `VERBOSE=1` argument:
+``make VERBOSE=1``
+For more comprehensive troubleshooting, use the Diagnostics Utility for
+Intel® oneAPI Toolkits, which provides system checks to find missing
+dependencies and permissions errors.
+[Learn more](https://software.intel.com/content/www/us/en/develop/documentation/diagnostic-utility-user-guide/top.html).
+
+ ### In Third-Party Integrated Development Environments (IDEs)
+
+You can compile and run this tutorial in the Eclipse* IDE (in Linux*) and the Visual Studio* IDE (in Windows*). For instructions, refer to the following link: [Intel® oneAPI DPC++ FPGA Workflows on Third-Party IDEs]([https://software.intel.com/en-us/articles/intel-oneapi-dpcpp-fpga-workflow-on-ide](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-oneapi-dpcpp-fpga-workflow-on-ide.html))
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/README.md:372]:
```suggestion
You can compile and run this tutorial in the Eclipse* IDE (in Linux*) and the Visual Studio* IDE (in Windows*). For instructions, refer to the following link: [Intel® oneAPI FPGA Workflows on Third-Party IDEs]([https://software.intel.com/en-us/articles/intel-oneapi-dpcpp-fpga-workflow-on-ide](https://www.intel.com/content/www/us/en/developer/articles/technical/intel-oneapi-dpcpp-fpga-workflow-on-ide.html))
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/src/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+endif()
+
+# This is for finding the install path to the necessary include files
+if(DEFINED ENV{INTELFPGAOCLSDKROOT})
+   set(SDK_ROOT_PATH $ENV{INTELFPGAOCLSDKROOT})
+else()
+   message(FATAL_ERROR "The INTELFPGAOCLSDKROOT environment variable must be defined for this tutorial to compile.")
+endif()
+
+# A DPC++ ahead-of-time (AoT) compile processes the device code in two stages.
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/src/CMakeLists.txt:42]:
```suggestion
# A SYCL ahead-of-time (AoT) compile processes the device code in two stages.
```

File: DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/src/CMakeLists.txt [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+set_target_properties(${EMULATOR_TARGET} PROPERTIES LINK_FLAGS "${EMULATOR_LINK_FLAGS}")
+add_custom_target(fpga_emu DEPENDS ${EMULATOR_TARGET})
+
+###############################################################################
+### Generate Report
+###############################################################################
+# To compile manually:
+#   icpx -fintelfpga -Xshardware -Xsboard=<FPGA_DEVICE> -fsycl-link=early hostpipes.cpp -o hostpipes_report.a
+set(FPGA_EARLY_IMAGE ${TARGET_NAME}_report.a)
+# The compile output is not an executable, but an intermediate compilation result unique to DPC++.
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/src/CMakeLists.txt:72]:
```suggestion
# The compile output is not an executable, but an intermediate compilation result unique to SYCL.
```

**Review by yuguen-intel — CHANGES_REQUESTED** [review]

File: DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/src/CMakeLists.txt [hunk 10]
```diff
@@ -0,0 +1,94 @@
+Please remove this 3-line comment
```
Comment (yuguen-intel) [inline DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/src/CMakeLists.txt:1]: I don't think that this did what you expected :)

**Review by yuguen-intel — CHANGES_REQUESTED** [review]

File: DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/src/CMakeLists.txt (hunk 10 above)
Comment (rho180) [reply DirectProgramming/DPC++FPGA/Tutorials/Features/experimental/hostpipes/src/CMakeLists.txt]: Done

**Pull request merged**